File: 2021/csharp/Aoc2021/Day11.cs
namespace Aoc2021;

public class Day11 : IDay
{
    public override string Day { get; } = nameof(Day11)[3..];

    public override string Result1()
    {
        return TotalFlashCount100(InputLines)
            .ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        return FirstStepFullFlash(InputLines)
            .ToString(CultureInfo.InvariantCulture);
    }

    public static int TotalFlashCount100(IEnumerable<string> lines)
    {
        int[][] grid = BuildGrid(lines);
        return FlashCount(grid).Take(100).Sum();
    }

    public static int FirstStepFullFlash(IEnumerable<string> lines)
    {
        int[][] grid = BuildGrid(lines);
        return FlashCount(grid).TakeWhile(n => n != 100).Count() + 1;
    }

    private static IEnumerable<int> FlashCount(int[][] grid)
    {
        while (true)
        {
            List<(int x, int y)> flashes =
                AllPoints()
                .Aggregate(
                    new List<(int x, int y)>(),
                    (flashes, p) =>
                    {
                        int power = ++grid[p.y][p.x];
                        if (power == 10)
                        {
                            flashes.Add((p.x, p.y));
                        }
                        return flashes;
                    }
                );

            yield return Flash(flashes, grid);

            _ = AllPoints()
                .Aggregate(
                    0,
                    (_, p) =>
                    {
                        if (grid[p.y][p.x] > 9)
                        {
                            grid[p.y][p.x] = 0;
                        }
                        return _;
                    }
                );
        }
    }

    private static int Flash(IEnumerable<(int x, int y)> points, int[][] grid)
    {
        int flashCount = 0;
        foreach ((int x, int y) in points)
        {
            List<(int x, int y)> flashes = new();
            foreach ((int x, int y) p in AdjacentPoints(x, y))
            {
                int power = ++grid[p.y][p.x];
                if (power == 10)
                {
                    flashes.Add(p);
                }
            }
            flashCount += 1 + Flash(flashes, grid);
        }
        return flashCount;
    }

    private const int N = 10;

    private static int[][] BuildGrid(IEnumerable<string> lines)
    {
        return lines
            .Select(line => line.Select(c => c - '0').ToArray())
            .ToArray();
    }

    private static IEnumerable<(int x, int y)> AllPoints()
    {
        for (int y = 0; y < N; y++)
        {
            for (int x = 0; x < N; x++)
            {
                yield return (x, y);
            }
        }
    }

    private static IEnumerable<(int x, int y)> AdjacentPoints(int x, int y)
    {
        return new (int x, int y)[]
            {
                (x - 1, y),
                (x + 1, y),
                (x    , y - 1),
                (x    , y + 1),
                (x - 1, y - 1),
                (x + 1, y + 1),
                (x + 1, y - 1),
                (x - 1, y + 1),
            }
            .Where(p => p.x >= 0 && p.x < N && p.y >= 0 && p.y < N);
    }
}

File: 2021/fsharp/Aoc2021/Day11.fs
module Day11

open Utils

type Point = int * int
type Grid = int array array

module Grid =
    let build (lines: string list) =
        lines
        |> List.map (fun line ->
            line.ToCharArray()
            |> Array.map (fun c -> int c - int '0'))
        |> List.toArray

    let contains (_grid: Grid) (x, y) =
        let nx, ny = 10, 10
        x >= 0 && x < nx && y >= 0 && y < ny

    let adjacentPoints (grid: Grid) (x, y) =
        [ x - 1, y
          x + 1, y
          x, y - 1
          x, y + 1
          x - 1, y - 1
          x + 1, y + 1
          x + 1, y - 1
          x - 1, y + 1 ]
        |> List.filter (grid |> contains)

    let flashCount (grid: Grid) =
        let incrementAll () =
            grid
            |> Array.indexed
            |> Array.fold
                (fun flashes (y, row) ->
                    row
                    |> Array.indexed
                    |> Array.fold
                        (fun flashes (x, power) ->
                            let power = power + 1
                            grid.[y].[x] <- power

                            if power = 10 then
                                (x, y) :: flashes
                            else
                                flashes)
                        flashes)
                List.empty

        let clearFlashed () =
            grid
            |> Array.iteri (fun y row ->
                row
                |> Array.iteri (fun x power ->
                    if power >= 10 then grid.[y].[x] <- 0
                    ())

                ())

        let flash flashes =
            let rec loop flashes count =
                match flashes with
                | [] -> count
                | p :: flashes' ->
                    let adjacentFlashes =
                        p
                        |> adjacentPoints grid
                        |> List.fold
                            (fun flashes (x, y) ->
                                let power = grid.[y].[x] + 1
                                grid.[y].[x] <- power

                                if power = 10 then
                                    (x, y) :: flashes
                                else
                                    flashes)
                            List.empty

                    loop (flashes' @ adjacentFlashes) (count + 1)

            loop flashes 0

        seq {
            while true do
                yield incrementAll () |> flash
                clearFlashed ()
        }

let totalFlashCount100 lines =
    lines
    |> Grid.build
    |> Grid.flashCount
    |> Seq.take 100
    |> Seq.sum

let firstStepFullFlash lines =
    lines
    |> Grid.build
    |> Grid.flashCount
    |> Seq.takeWhile (fun count -> count <> 100)
    |> Seq.length
    |> (+) 1

let input = "11" |> readInputLines |> Seq.toList

let job1 () = input |> totalFlashCount100 |> string

let job2 () = input |> firstStepFullFlash |> string
